### PR TITLE
word wrap utility

### PR DESCRIPTION
**Pull request by Lupus590**
~~I'm also making a cc.pretty version, it will be a separate commit on this PR.~~

### CODE REVIEW

**Comment by Lupus590** [comment]
Tests on the pretty version are not passing on my machine, I can't figure out why.

File: src/main/resources/data/computercraft/lua/rom/apis/textutils.lua [hunk 1]
```diff
@@ -822,3 +822,74 @@ function complete(sSearchText, tSearchTable)
...
+        end
+    end
+
+    push_line()
+
+    -- Normalise the strings suitable for use with blit. We could skip this and
+    -- just use term.write, but saves us a clearLine call.
+    for k, line in pairs(lines) do
+        line = line:sub(1, width)
+        if #line < width then line = line .. (" "):rep(width - #line) end
```
Comment (SquidDev) [inline src/main/resources/data/computercraft/lua/rom/apis/textutils.lua:890]: I think we should drop this line. If people want to pad it afterwards that's fine, but the original code was written for a very specific purpose, and I don't think this is as general.

**Review by SquidDev — CHANGES_REQUESTED** [review]
Well, the build failures due to trunk are moderately annoying. Even forcing a rebuild didn't fix it.

Rather than chucking it in textutils (eww, globals) or `cc.pretty` (not really related to this at all), I wonder if it's worth chucking it into a new module. Not sure about names - `cc.textutils` is probably confusing, so maybe `cc.strings`?

File: src/main/resources/data/computercraft/lua/rom/apis/textutils.lua (hunk 1 above)
Comment (Lupus590) [reply src/main/resources/data/computercraft/lua/rom/apis/textutils.lua]: I moved this to a new function in the new module.

File: src/test/resources/test-rom/spec/modules/cc/strings_spec.lua [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        it("wraps lines", function()
+            expect(str.wrap("test string is long")[1]):eq("test string is long")
+
+            expect(str.wrap("test string is long", 15)[1]):eq("test string is ")
+            expect(str.wrap("test string is long", 15)[2]):eq("long")
+
+            expect(str.wrap("test string is long", 12)[1]):eq("test string ")
+            expect(str.wrap("test string is long", 12)[2]):eq("is long")
+
+            expect(str.wrap("test string is long", 11)[1]):eq("test string")
```
Comment (Lupus590) [inline src/test/resources/test-rom/spec/modules/cc/strings_spec.lua:23]: This test fails, I can't figure out how to remove the trailing space nicely.

Comment (SquidDev) [reply src/test/resources/test-rom/spec/modules/cc/strings_spec.lua:23]: We only take the length of the line into account when appending new words, not spaces. So `test string` is fine, then we see a space (`test string `) and _then_ we add a new line.

The normalisation code above (`:sub(1, width)`) should fix that problem, hence me suggesting keeping that but not the padding with additional spaces :p.

File: src/main/resources/data/computercraft/lua/rom/modules/main/cc/strings.lua [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+-- This is done by adding spaces or truncating the string.
+--
+-- @tparam string line The string to normalise.
+--
+-- @tparam[opt] number width The width to contrain to, defaults to the width of
+-- the terminal.
+--
+-- @treturn { string... } The wrapped input string.
+-- @usage strings.normalise( "long string", 5 )
+-- @usage strings.normalize( "long string", 5 )
```
Comment (SquidDev) [inline src/main/resources/data/computercraft/lua/rom/modules/main/cc/strings.lua:87]: These two examples are the same!

**Review by SquidDev — CHANGES_REQUESTED** [review]
Sorry, I'm very much nitpicking here!

File: src/main/resources/data/computercraft/lua/rom/modules/main/cc/strings.lua [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+local function normalise(line, width)
```
[the remark this answers is not in the document]
Comment (Lupus590) [reply src/main/resources/data/computercraft/lua/rom/modules/main/cc/strings.lua:88]: It doesn't just pad though, it also truncates.

File: src/main/resources/data/computercraft/lua/rom/modules/main/cc/strings.lua (hunk 3 above)
Comment (Lupus590) [reply src/main/resources/data/computercraft/lua/rom/modules/main/cc/strings.lua:87]: normali**s**e vs normali**z**e

Comment (Lupus590) [reply src/main/resources/data/computercraft/lua/rom/modules/main/cc/strings.lua:87]: wow, even bolding the s and z doesn't make that ovious

File: src/main/resources/data/computercraft/lua/rom/modules/main/cc/strings.lua [hunk 5]
```diff
@@ -0,0 +1,105 @@
+--- Varous string utilities that might go into textutils if we weren't avoiding
+-- globals.
+
+local expect = require "cc.expect".expect
+
+--- Converts a long string into a table of strings of length shorter than a
+-- given width.
+--
+-- It attemps to split on whitespace and converts \n into the next table entry.
```
[the remark this answers is not in the document]
Comment (Lupus590) [reply src/main/resources/data/computercraft/lua/rom/modules/main/cc/strings.lua]: Is the new version better?

File: src/main/resources/data/computercraft/lua/rom/modules/main/cc/strings.lua (hunk 4 above)
[the remark this answers is not in the document]
Comment (SquidDev) [reply src/main/resources/data/computercraft/lua/rom/modules/main/cc/strings.lua:89]: Yeah, gah. `fix_width`, `ensure_width` maybe?

File: src/main/resources/data/computercraft/lua/rom/modules/main/cc/strings.lua (hunk 3 above)
Comment (SquidDev) [reply src/main/resources/data/computercraft/lua/rom/modules/main/cc/strings.lua:88]: Fine, under alpha-renaming the two examples are the same! :p. 

I guess my point is that the second doesn't divulge any new information about where the function can be used, and so is that sense isn't very useful.

**Comment by SquidDev** [comment]
Merged manually in 16d74dd2e8709b4df35a826ca053b4a3c332f393. I thought I could push to PR branches, but either something is wrong with my setup or GH removed that feature.

**Pull request closed without merging**